libmount/src/tab_parse.c

```diff
@@ -673,7 +681,25 @@ static int kernel_fs_postparse(struct libmnt_table *tb,
 	if (src && strcmp(src, "/dev/root") == 0) {
 		char *real = NULL;
 
-		rc = mnt_guess_system_root(fs->devno, tb->cache, &real);
+		/* We will only call mnt_guess_system_root() if it has not
+		 * been called before. Inside a container, mountinfo can contain
+		 * many lines with /dev/root.
+		 */
+		if (saveguess->rc == 0 && saveguess->real == NULL) {
+			rc = mnt_guess_system_root(fs->devno, tb->cache, &real);
+			if (rc == 0 && real != NULL)
+				saveguess->real = strdup(real);
+			saveguess->rc = rc;
+		} else {
+			/* This means that we already have run the
+			 * mnt_guess_system_root() and that we want to reuse the
+			 * old result
+			 */
+			rc = saveguess->rc;
+			if (rc == 0 && saveguess->real != NULL)
+				real = strdup(saveguess->real);
```

Comment:
Please, always check strdup() result, and `return -ENOMEM` is a case of NULL.

Reply:
My bad. Somehow I thought that the later code would deal "gracefully"  with a NULL pointer. I have corrected this now.